src/linux/build_id_reader.rs

```diff
@@ -0,0 +1,225 @@
+use crate::errors::BuildIdReaderError as Error;
+use crate::minidump_format::GUID;
+use goblin::{
+    container::{Container, Ctx, Endian},
+    elf,
+};
+
+const NOTE_SECTION_NAME: &[u8] = b".note.gnu.build-id\0";
+
+pub trait ModuleMemory {
+    type Memory: std::ops::Deref<Target = [u8]>;
+
+    fn read_module_memory(&self, offset: u64, length: u64) -> std::io::Result<Self::Memory>;
+}
+
+impl<'a> ModuleMemory for &'a [u8] {
+    type Memory = Self;
+
+    fn read_module_memory(&self, offset: u64, length: u64) -> std::io::Result<Self::Memory> {
+        Ok(self
+            .get(offset as usize..(offset + length) as usize)
+            .unwrap())
+    }
+}
+
+fn read<T: ModuleMemory>(mem: &T, offset: u64, length: u64) -> Result<T::Memory, Error> {
+    mem.read_module_memory(offset, length)
+        .map_err(|error| Error::ReadModuleMemory {
+            offset,
+            length,
+            error,
+        })
+}
+
+pub fn read_build_id(module_memory: impl ModuleMemory) -> Result<Vec<u8>, Error> {
+    let reader = ElfBuildIdReader::new(module_memory)?;
+    let program_headers = match reader.read_from_program_headers() {
+        Ok(v) => return Ok(v),
+        Err(e) => Box::new(e),
+    };
+    let section = match reader.read_from_section() {
+        Ok(v) => return Ok(v),
+        Err(e) => Box::new(e),
+    };
+    let generated = match reader.generate_from_text() {
+        Ok(v) => return Ok(v),
+        Err(e) => Box::new(e),
+    };
+    Err(Error::Aggregate {
+        program_headers,
+        section,
+        generated,
+    })
+}
+
+pub struct ElfBuildIdReader<T> {
+    module_memory: T,
+    header: elf::Header,
+    context: Ctx,
+}
+
+impl<T: ModuleMemory> ElfBuildIdReader<T> {
+    pub fn new(module_memory: T) -> Result<Self, Error> {
+        // We could use `Ctx::default()` (which defaults to the native system), however to be extra
+        // permissive we'll just use a 64-bit ("Big") context which would result in the largest
+        // possible header size.
```

Comment:
The only not-completely-contrived case I could think of is the minidump-writer process being 64-bit and the child process being 32-bit, or vice versa. But I think it _is_ very unlikely, especially nowadays.

Reply:
Ah yes, because we're reading it from `/proc/<pid>/maps` so we need to strip the suffix for deleted files (or it'd mess up with symbol resolution later when we process the minidump). Somehow my mind had conflated the fact that we also want to read the **SONAME** from memory with this patch, but that's actually issue #79 which is the natural follow-up to this.